From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lukas81298 <lukas81298@gommehd.net>
Date: Fri, 30 Apr 2021 23:59:59 +0200
Subject: [PATCH] updated to latest paper


diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000000000000000000000000000000000000..db846856f48f81b05a6813b887913c4247d1f9b5
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,3 @@
+target/
+*.iml
+.idea/
\ No newline at end of file
